### ISSUE TITLE
jit-compilable "apply mask"-operation

### ISSUE THREAD

**Issue by SimiPixel** [google/jax#9579]
Working with masks is quite frustrating in jit-compiled functions and i would be amazing to have a function such that

```python
@jax.jit
def f(x, mask):
    assert x.shape == mask.shape
    # where mask is a boolean-array
    return jnp.apply_mask(x, mask)
```

this would be possible. This is the underlying problem why i require this.

```python
from jax import random 
import jax
import jax.numpy as jnp

def biject_alpha(idx, val):
    alpha, i, key = val
    # hyperparameters for randomisation
    cdf_bins=20

    key, consume = random.split(key)
    samples = random.uniform(consume, (cdf_bins-1,), maxval=1.0)
    samples = jnp.hstack((jnp.array([0.0]), samples))
    montonous = jnp.cumsum(samples)
    cdf = montonous/montonous[-1]

    # idx of first occurenece
    mask = (i-1) == idx
    start_idx = jnp.select(mask, jnp.arange(len(mask)))

    # biject slice
    alpha_i = alpha[mask]
    bijected_alpha = cdf[(alpha_i//(1/cdf_bins)).astype(int)]
    alpha = jax.lax.dynamic_update_slice(alpha, bijected_alpha, start_idx)
    return (alpha, i, key)

def cosInterpolateRandomized(x, xp, fp, consume):
    i = jnp.clip(jnp.searchsorted(xp, x, side='right'), 1, len(xp) - 1)
    dx = xp[i] - xp[i - 1]
    alpha = (x - xp[i - 1])/dx 
    alpha, _, _ = jax.lax.fori_loop(0, len(xp)-1, biject_alpha, (alpha, i, consume))

    def cos_interpolate(x1, x2, alpha):
        """ x2 > x1
        """
        return (x1+x2)/2 + (x1-x2)/2 * jnp.cos(alpha*jnp.pi)

    f = jnp.where((dx == 0), fp[i], jax.vmap(cos_interpolate)(fp[i-1], fp[i], alpha))
    f = jnp.where(x > xp[-1], fp[-1], f)
    return f 
```

**Comment by jakevdp** [comment]
Hi - can you give an example of what you want the `apply_mask` function to do?

**Comment by SimiPixel** [comment]
Hey,

thanks for the super quick reply.
So, i tried to clear it up in my post. In my example use-case, there is on line in the `biject_alpha`-function that i do not know how to solve differently.

**Comment by jakevdp** [comment]
Ah, I see, sorry. I was looking for a definition of `apply_mask`, and didn't look at the comments near its use.

As a general answer, there is currently no JIT-compatible way to execute `y = x[mask]` for a dynamic `mask`. This is because the size of `y` must be known statically (at trace/compile time).

But you may be able to solve this for the particular case you're interested in... I'm having trouble understanding the code, though, because there's no indication of the types/sizes of the inputs to `cosInterpolateRandomized`.

**Comment by SimiPixel** [comment]
I added some asserts that hopefully make this more clear.
The function is conceptually identical to `jax.numpy.interp` [https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.interp.html](url)

**Comment by SimiPixel** [comment]
Actually, this `alpha_i` is deceiving. Let me try to clarify:

```python
x = jnp.arange(5)
xp = jnp.array([0, 1.5, 3, 4])
i = jnp.clip(jnp.searchsorted(xp, x, side="right"), 1, len(xp)-1)
i
[1,1,2,3]
```
Then,
```python
dx = xp[i] - xp[i-1]
alpha = (x - xp[i-1])/dx
alpha
[0., 0.66667, 0.33334, 0., 1.]
```
Then,
```
biject_alpha(0, (alpha, i, random.PRNGKey(0)))
[0., 0.83654, 0.33334, 0., 1.]
```

I hope this helps :)

**Comment by SimiPixel** [comment]
This seems to do the trick, it is probably a nicer solution anyway. Though i do not fully understand why this masking operation know is fine.

```python
from jax import random 
import jax
import jax.numpy as jnp

def biject_alpha(alpha, cdf):
    cdf_bins = len(cdf)
    return cdf[(alpha//(1/cdf_bins)).astype(int)]
    #return jax.lax.dynamic_index_in_dim(cdf, (alpha//(1/cdf_bins)).astype(int))

def generate_cdf(key):
    cdf_bins=20
    samples = random.uniform(key, (cdf_bins-1,), maxval=1.0)
    samples = jnp.hstack((jnp.array([0.0]), samples))
    montonous = jnp.cumsum(samples)
    cdf = montonous/montonous[-1]
    return cdf

@jax.jit
def cosInterpolateRandomized(x, xp, fp, key):
    i = jnp.clip(jnp.searchsorted(xp, x, side='right'), 1, len(xp) - 1)
    dx = xp[i] - xp[i - 1]
    alpha = (x - xp[i - 1])/dx 

    key, *consume = random.split(key, len(xp)+1)
    consume = jnp.array(consume).reshape((len(xp), 2))
    consume = consume[i - 1]
    cdfs = jax.vmap(generate_cdf)(consume)
    alpha = jax.vmap(biject_alpha)(alpha, cdfs)

    def cos_interpolate(x1, x2, alpha):
        """ x2 > x1
        """
        return (x1+x2)/2 + (x1-x2)/2 * jnp.cos(alpha*jnp.pi)

    f = jnp.where((dx == 0), fp[i], jax.vmap(cos_interpolate)(fp[i-1], fp[i], alpha))
    f = jnp.where(x > xp[-1], fp[-1], f)
    return f

x = jnp.arange(5)
xp = jnp.array([0, 1.5, 3, 4])
fp = jnp.array([0, 3, 6, 8.])
cosInterpolateRandomized(x, xp, fp, random.PRNGKey(0))
[0, 2.46, 3.62, 6, 8]
```

**Comment by smorad** [comment]
Masking is also pretty important for dealing with variable-length timeseries (transformers, RNNs, etc). Are `jax` transformers just never jitted?

**Comment by mattjj** [comment]
They're jitted! But if people need to batch together variable length sequences, they write the masking and padding manually.

**Comment by mohamad-amin** [comment]
Could you please point me to an example of this manual masking and padding?

**Comment by jakevdp** [comment]
> it needs it to keep the original order in case the values being sorted are equal

`argsort` in JAX is always a stable sort in JAX, so you should be able to depend on this being true

> If anyone has a better idea of doing this, I'll appreciate them sharing.

This version follows your first algorithm, ensuring shapes are static so it will remain compatible with `jit` and other jax transformations

```python
import jax
import jax.numpy as jnp
from functools import partial

@jax.jit
def f(x, y):
  assert x.shape == y.shape
  B, C = x.shape
  indices = jnp.argmax(y, axis=1)
  c = x[jnp.arange(B), indices][:, None]
  w = jax.vmap(partial(jnp.delete, assume_unique_indices=True))(x, indices)
  return jax.nn.softplus(jax.scipy.special.logsumexp(w - c, axis=1)).mean()
```
